In [1]:
import pandas as pd
import numpy as np
import re

### Excercise 1: Prosecco title length

<b>The hypothesis:</b> <br>"Manufacturers of poor Prosecco wines attempt to compensate for the low quality of their wines by choosing long and fancy names."

In [2]:
# Loading dataset
wine_df = pd.read_csv('../Datasets/winemag-data-130k-v2.csv')

# Creating a new dataframe with Prosecco wines only
prosecco_df = wine_df.loc[wine_df['variety'] == 'Prosecco']
prosecco_df.reset_index(inplace=True)
prosecco_df = prosecco_df[['title', 'price', 'points']]
prosecco_df

title  price  points
0    Bellussi NV Extra Dry  (Prosecco di Valdobbiad...   15.0      86
1      Paladin 2007 Millesimato Brut Prosecco (Veneto)   20.0      86
2    Perlage 2008 Col di Manza Extra Dry Millesimat...   22.0      86
3    Sant Eurosia 2007 Brut  (Prosecco di Valdobbia...   16.0      86
4    Sant Eurosia 2007 Millesimato Dry  (Prosecco d...   18.0      86
..                                                 ...    ...     ...
231            Mionetto NV Brut  (Prosecco del Veneto)   12.0      85
232   Moletto NV Frizzante Prosecco (Marca Trevigiana)   14.0      85
233  Perlage NV Canah Brut  (Prosecco di Valdobbiad...   19.0      85
234          Valdo NV Extra Dry  (Prosecco del Veneto)   10.0      85
235  Col Vetoraz Spumanti NV  Prosecco Superiore di...   38.0      91

[236 rows x 3 columns]

In [3]:
# Adding a new column with title length 
for row in prosecco_df.iterrows():
    index = row[0]
    title = row[1][0]
    title_length = len(title)
    prosecco_df.loc[index, 'title_length'] = title_length

prosecco_df['title_length'] = prosecco_df['title_length'].astype(int)


# Creating two new dataframes: 
# One for good prosecco (rating > 89) and one for bad prosecco (rating < 85)
good_prosecco = prosecco_df.loc[prosecco_df['points'] > 89].sort_values(['points'], ascending=False)
bad_prosecco = prosecco_df.loc[prosecco_df['points'] < 85].sort_values(['points'], ascending=False)
good_prosecco.head(20)

title  price  points  \
235  Col Vetoraz Spumanti NV  Prosecco Superiore di...   38.0      91   
123  Ruggeri & C. 2007 Giustino B. Extra Dry  (Pros...   36.0      91   
162           Adami NV  Prosecco Superiore di Cartizze   32.0      90   
216  Bisol NV Cartizze  (Prosecco Superiore di Cart...   41.0      90   
215     Bortolomiol NV  Prosecco Superiore di Cartizze   30.0      90   
180  Bortolomiol 2006 Cartizze  (Prosecco Superiore...   35.0      90   
179  Astoria NV Cartizze  (Prosecco Superiore di Ca...   21.0      90   
178  Astoria 2006 Millesimato Extra Dry  (Prosecco ...   20.0      90   
163        Bisol NV Crede  (Prosecco di Valdobbiadene)   25.0      90   
39   Sorelle Bronca NV Extra Dry Particella 68  (Pr...    NaN      90   
103  Bortolomiol 2008 Cartizze Dry  (Prosecco Super...   30.0      90   
145  Sorelle Bronca NV Particella 68 Extra Dry  (Pr...   24.0      90   
144  Adami NV Cartizze Dry  (Prosecco Superiore di ...   32.0      90   
143  Bisol 2007 Cartizze  (Prosecco Superiore di Ca...   48.0      90   
124    Ruggeri & C. NV  Prosecco Superiore di Cartizze   30.0      90   
104  Nino Franco 2007 Rive di San Floriano Brut  (P...   30.0      90   
154  Bortolomiol 2006 Cartizze  (Prosecco Superiore...   35.0      90   

     title_length  
235            55  
123            68  
162            40  
216            51  
215            46  
180            59  
179            53  
178            76  
163            43  
39             70  
103            63  
145            70  
144            55  
143            53  
124            47  
104            71  
154            59

In [4]:
bad_prosecco.head(20)

title  price  points  \
134  Bellenda NV San Fermo Brut  (Prosecco di Coneg...   19.0      84   
93   Cantina San Martino NV Pittaro Extra Dry  (Pro...   15.0      84   
94   Lisabella NV Gran Resèe Prosecco (Colli Trevig...   12.0      84   
95   Mionetto NV Certified Organic Extra Dry Prosec...   16.0      84   
117               Tiamo NV Extra Dry Prosecco (Veneto)   14.0      84   
214  Collalbrigo NV Extra Dry  (Prosecco di Conegli...   16.0      84   
213     Canella NV Extra Dry  (Prosecco di Conegliano)   18.0      84   
131  Terra Serena NV Extra Dry  (Prosecco di Conegl...   12.0      84   
132  Col Saliz NV Extra Dry  (Prosecco di Valdobbia...   15.0      84   
133       Varaschin NV  Prosecco Superiore di Cartizze   29.0      84   
212  Tenuta Santomè NV Extra Dry  (Prosecco del Ven...    NaN      84   
146  Perlage NV Col di Manza Extra Dry  (Prosecco d...   22.0      84   
211  Villa Granda NV Frizzante Prosecco (Colli Trev...    NaN      84   
148  Toffoli NV Brut  (Prosecco di Conegliano e Val...   18.0      84   
149                          Zonin NV Prosecco (Italy)   12.0      84   
161  Carmina NV Brut  (Prosecco di Conegliano e Val...   15.0      83   
32                           Tosti NV Prosecco (Italy)   15.0      83   
147     Varaschin NV Brut  (Prosecco di Valdobbiadene)   19.0      83   
164  Carmina NV Brut  (Prosecco di Conegliano e Val...   16.0      82   
129  Maschio dei Cavalieri NV Maschio dei Cavalieri...   20.0      82   

     title_length  
134            68  
93             63  
94             51  
95             57  
117            36  
214            50  
213            46  
131            67  
132            51  
133            44  
212            50  
146            62  
211            53  
148            57  
149            25  
161            57  
32             25  
147            46  
164            57  
129            75

In [5]:
# Calculating the mean title length for good and bad prosecco wines
good_prosecco_len_sum = 0
for row in good_prosecco.iterrows():
    index = row[0]
    title_length = row[1][3]
    good_prosecco_len_sum += title_length

good_prosecco_len_avg = good_prosecco_len_sum / len(good_prosecco)
print(f'Good prosecco mean title length: {good_prosecco_len_avg}')

bad_prosecco_len_sum = 0
for row in bad_prosecco.iterrows():
    index = row[0]
    title_length = row[1][3]
    bad_prosecco_len_sum += title_length

bad_prosecco_len_avg = bad_prosecco_len_sum / len(bad_prosecco)
print(f'Bad prosecco mean title length: {bad_prosecco_len_avg}')

Good prosecco mean title length: 57.588235294117645
Bad prosecco mean title length: 53.73913043478261


<b>Conclusion:</b> <br>It seems that the hypothesis was wrong, as the data indicates the opposite: That better Proseccos have longer names.  <br><br>

### Excercise 2: Ramen reviews 
<i> Which country has the best ramen?</i>

In [6]:
# Loading dataset
ramen = pd.read_csv('../Datasets/ramen-ratings.csv')
ramen['Stars'].replace(to_replace='Unrated', value=0, inplace=True)
ramen['Stars'] = ramen['Stars'].astype(float)

# Cleaning dataset. Replacing null values for ratings (Unrated) with 0. Set 'Stars' to float
ramen.drop(labels='Review #', axis=1, inplace=True)
ramen['Stars'].replace(to_replace='Unrated', value=0, inplace=True)
ramen['Stars'] = ramen['Stars'].astype(float)

# Grouping data by country and aggregating the mean and the 10th and 90th quantiles
ramen_country_stats = ramen.groupby(by=['Country'], sort=True).agg([np.mean])
ramen_country_stats['q10'] = ramen.groupby(by=['Country'], sort=True).describe(percentiles=[0.1])[('Stars',   '10%')]
ramen_country_stats['q90'] = ramen.groupby(by=['Country'], sort=True).describe(percentiles=[0.9])[('Stars',   '90%')]

ramen_country_stats['Count'] = ramen.groupby(by=['Country'], sort=True).describe()[('Stars', 'count')]
ramen_country_stats.sort_values(by=[('Stars', 'mean'), 'q90'], axis=0, ascending=False)

Stars    q10    q90  Count
                   mean                     
Country                                     
Brazil         4.350000  4.000  4.800    5.0
Sarawak        4.333333  4.000  4.800    3.0
Cambodia       4.200000  3.500  5.000    5.0
Malaysia       4.127564  3.125  5.000  156.0
Singapore      4.126147  3.200  5.000  109.0
Indonesia      4.067460  3.250  5.000  126.0
Japan          3.981605  3.000  5.000  352.0
Myanmar        3.946429  2.975  5.000   14.0
Fiji           3.875000  3.475  4.175    4.0
Hong Kong      3.801825  2.750  5.000  137.0
South Korea    3.766019  2.750  5.000  309.0
United States  3.750000  3.750  3.750    1.0
Mexico         3.730000  3.000  4.000   25.0
Bangladesh     3.714286  3.250  4.000    7.0
Taiwan         3.665402  2.000  5.000  224.0
Germany        3.638889  3.000  4.350   27.0
Poland         3.625000  3.250  4.000    4.0
Hungary        3.611111  2.950  4.150    9.0
Dubai          3.583333  3.350  3.750    3.0
Finland        3.583333  3.500  3.700    3.0
Holland        3.562500  3.500  3.675    4.0
Nepal          3.553571  3.325  4.175   14.0
Estonia        3.500000  3.300  3.700    2.0
Ghana          3.500000  3.500  3.500    2.0
USA            3.457043  2.000  4.750  323.0
China          3.421893  1.750  4.500  169.0
India          3.395161  2.000  4.250   31.0
Thailand       3.384817  2.000  4.750  191.0
Philippines    3.329787  1.900  4.500   47.0
Colombia       3.291667  2.875  3.625    6.0
Sweden         3.250000  3.050  3.450    3.0
Vietnam        3.187963  2.000  4.000  108.0
Australia      3.138636  2.025  4.000   22.0
Pakistan       3.000000  2.400  3.600    9.0
UK             2.997101  1.500  4.050   69.0
Netherlands    2.483333  0.500  3.500   15.0
Canada         2.243902  0.250  3.500   41.0
Nigeria        1.500000  1.500  1.500    1.0

<b>Conclusion:</b> On average, Brazil has the best ramen with a mean rating of 4.35. They also have a very high 10th percentile, so if you get a hold of Brazilian ramen, you know that it's going to be good. However, the entire country only sells 5 different ramen products with a 90th percentile of 4.8, while Malaysia sells more than 150 different ramen products. Malaysia has a slightly lower average rating of 4.13, but has a higher 90th percentile of 5.00. I would therefore recommend Cambodian ramen on a general basis, because it will most likely be easier to find in stores than Brazilian ramen, while at the same time supplying a wider range of top-of-the-line ramen products. 

<b><i>But which style of ramen is most popular in each country?</i></b>

In [7]:
# Loading a fresh dataset
ramen = pd.read_csv('../Datasets/ramen-ratings.csv')

# Counting the number of different ramen products in each country, as well as the number of different styles 
ramen_country_style = ramen.groupby(by=['Country', 'Style']).count().drop(['Brand', 'Variety', 'Stars', 'Top Ten'], axis=1)
ramen_country_totals = ramen.groupby(by=['Country']).count().drop(['Style', 'Brand', 'Variety', 'Stars', 'Top Ten'], axis=1)

# Calculating the proportion for each different style in each country
for row1 in ramen_country_style.iterrows():
    index = row1[0]
    country = row1[0][0]
    review_count = float(row1[1][0])

    for row2 in ramen_country_totals.iterrows():
        total = float(row2[1][0])
        totals_country = row2[0]

        if country == totals_country:
            ramen_country_style.loc[index, '%'] = review_count / total

# Dropping the Review # column as we don't care about review count
ramen_country_style.drop(['Review #'], axis=1, inplace=True)

# Displaying the result
ramen_country_style

%
Country       Style          
Australia     Cup    0.772727
              Pack   0.227273
Bangladesh    Pack   1.000000
Brazil        Cup    0.400000
              Pack   0.600000
...                       ...
United States Pack   1.000000
Vietnam       Bowl   0.185185
              Cup    0.074074
              Pack   0.722222
              Tray   0.018519

[87 rows x 1 columns]

### Excercise 3: Customer data: Men vs Women

<b> Hypothesis: </b> "Men order more than women at restaurants. "

In [8]:
# Loading data
orders = pd.read_csv('../Datasets/orders.csv', low_memory=False)
customers = pd.read_csv('../Datasets/customers.csv')

# Preparing datasets for join operation. 
customers.rename(columns={'akeed_customer_id': 'customer_id'}, inplace=True)
customers = customers.loc[:, ['customer_id', 'gender']]
orders = orders.loc[:, ['customer_id', 'item_count']]
# The spelling of the genders is not consistent. Fixing that with regex
customers['gender'] = customers.gender.str.replace('(F|f)emale.*', 'Female', regex=True)
customers['gender'] = customers.gender.str.replace('^(M|m)ale.*', 'Male', regex=True)
customers['gender'] = customers.gender.str.replace('\s+', '<NA>', regex=True)

# Perform join so that we can calculate the average item count for each gender
merged = orders.join(customers.set_index('customer_id'), on='customer_id')
merged.drop('customer_id', axis=1, inplace=True)

merged.replace(to_replace='<NA>', value=pd.NA, regex=True, inplace=True)
print('Missing data:\n' , merged.isnull().sum() / merged.count().sum())
# Removing missing values, as we cannot use this data for the calculation. 
# 13% of the gender data is messing, but as the data most likely is missing at random, the impact should be negligible. 
merged.dropna(inplace=True)

# Aggregate the mean item count for each gender
merged_grouped = merged.groupby('gender').agg(np.mean)
merged_grouped.head(50)

Missing data:
 item_count    0.029682
gender        0.131905
dtype: float64


item_count
gender            
Female    2.381009
Male      2.414087

<b>Conclusion:</b> The hypothesis was wrong. Men and women order about the same number of items. <br><br>

### Excercise 4: Track billboard data from the glorious year 2000
Out of all the artists that reached the billboards in the year 2000, which artist held the highest average billboard rank?

In [9]:
# Loading dataset
billboard = pd.read_csv('../Datasets/billboard.csv')

# The column headers are values, not variable names. Tidying dataframe
# Adding all week values to a single list variable
weeks_listed = billboard.columns.tolist()[7:]
melted = pd.melt(billboard, id_vars=["artist.inverted", "track", 'time', 'genre', 'date.entered', 'date.peaked'],
                 value_vars=weeks_listed, var_name="Week", value_name="Rank")

melted.head()

artist.inverted                                  track  time genre  \
0      Destiny's Child               Independent Women Part I  3:38  Rock   
1              Santana                           Maria, Maria  4:18  Rock   
2        Savage Garden                     I Knew I Loved You  4:07  Rock   
3              Madonna                                  Music  3:45  Rock   
4  Aguilera, Christina  Come On Over Baby (All I Want Is You)  3:38  Rock   

  date.entered date.peaked       Week  Rank  
0   2000-09-23  2000-11-18  x1st.week  78.0  
1   2000-02-12  2000-04-08  x1st.week  15.0  
2   1999-10-23  2000-01-29  x1st.week  71.0  
3   2000-08-12  2000-09-16  x1st.week  41.0  
4   2000-08-05  2000-10-14  x1st.week  57.0

In [10]:
# Converting week variables from messy strings to integers
for row in melted.iterrows():
    index = row[0]
    week_str = row[1][6]
    week_int = int(re.sub('[^0-9]', '', week_str))
    melted.loc[index, 'Week'] = week_int

melted.head()

artist.inverted                                  track  time genre  \
0      Destiny's Child               Independent Women Part I  3:38  Rock   
1              Santana                           Maria, Maria  4:18  Rock   
2        Savage Garden                     I Knew I Loved You  4:07  Rock   
3              Madonna                                  Music  3:45  Rock   
4  Aguilera, Christina  Come On Over Baby (All I Want Is You)  3:38  Rock   

  date.entered date.peaked Week  Rank  
0   2000-09-23  2000-11-18    1  78.0  
1   2000-02-12  2000-04-08    1  15.0  
2   1999-10-23  2000-01-29    1  71.0  
3   2000-08-12  2000-09-16    1  41.0  
4   2000-08-05  2000-10-14    1  57.0

In [11]:
# Grouping the rows by "artist.inverted" and "track", and computing the average rank for each artist,
# Then sorting the resulting dataframe from lowest to highest average rank and displaying the top 10 artists
stats = melted.groupby(['artist.inverted', 'track']).agg(np.mean).sort_values(by='Rank', ascending=True)
stats.head(10)

Rank
artist.inverted              track                              
Santana                      Maria, Maria              10.500000
Madonna                      Music                     13.458333
N'Sync                       Bye Bye Bye               14.260870
Elliott, Missy "Misdemeanor" Hot Boyz                  14.333333
Destiny's Child              Independent Women Part I  14.821429
Iglesias, Enrique            Be With You               15.850000
Aaliyah                      Try Again                 16.656250
Savage Garden                I Knew I Loved You        17.363636
Houston, Whitney             My Love Is Your Love      17.857143
Pink                         There U Go                18.625000

<b>Conclusion:</b> Out of all the great artists of the year 2000, Santana held the highest average billboard rank, followed by Madonna and N'Sync! 